Fix race in MapKey

ralfonso opened the pull request:
Fix race in `ManualClock`
Fix race in `MapKey`

MapKey was previously exported in https://github.com/Netflix/spectator-go/pull/42

FWIW, the current MapKey behavior could done more safely by introducing a lock rather than precomputing during construction, as introduced in this change. I chose to avoid the lock on the assumption that users are not depending on the JIT computation behavior.

Make MapKey() thread-safe by precomputing the value during construction.
This does not change any documented behavior of Id, which is immutable
(apart from one case mentioned below). However, the lack of map clone in
Tags() means that previously a caller could construct a new Id, calls
Tags(), modify the resulting map, and then call MapKey(). This would
result in a different output after this change, which captures the tags
at construction.

I believe this is more aligned with the intent of this type. The
documented behavior mentions the lack of clone to warn the user of
modifying the map from Tags(), but does not prevent it.

Interestingly, in the previous implementation, this was likely hidden by the Registry calling 
MapKey(), which uses a mutex for registration via a blocking call to NewTimer() or friends. I suspect
that the most common use of Id has been to construct and immediately
register. Since the value for MapKey() was previously precomputed
lazily, subsequent calls after registration were threadsafe.

The use-case that led to this discovery is relying on `MapKey()` to populate a local cache of `Id` values for use by `PersistentTimers`. The cache was introduced to avoid the relatively expensive construction of a `PersistentTimer` (and its underlying side effects) that may already have been created. 



Race example for `ManualClock`
```➜ go test -race ./...
DEBUG: 2021/03/01 12:54:54 posting data to http://127.0.0.1:38093, payload 2 bytes
DEBUG: 2021/03/01 12:54:54 response HTTP 200:
DEBUG: 2021/03/01 12:54:54 posting data to http://127.0.0.1:36927, payload 2 bytes
INFO: 2021/03/01 12:54:54 Timed out attempting to POST to http://127.0.0.1:36927
==================
WARNING: DATA RACE
Read at 0x00c000280330 by goroutine 67:
  github.com/Netflix/spectator-go.(*ManualClock).Nanos()
      /home/r2/code/ext/spectator-go/clock.go:41 +0x3a
  github.com/Netflix/spectator-go.(*LogEntry).Log()
      /home/r2/code/ext/spectator-go/log_entry.go:54 +0x6c
  github.com/Netflix/spectator-go.(*HttpClient).doHttpPost()
      /home/r2/code/ext/spectator-go/http_client.go:135 +0x74c
  github.com/Netflix/spectator-go.(*HttpClient).PostJson()
      /home/r2/code/ext/spectator-go/http_client.go:143 +0x145
  github.com/Netflix/spectator-go.TestHttpClient_PostJsonTimeout()
      /home/r2/code/ext/spectator-go/http_client_test.go:124 +0xa6e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202

Previous write at 0x00c000280330 by goroutine 73:
  github.com/Netflix/spectator-go.(*ManualClock).SetFromDuration()
      /home/r2/code/ext/spectator-go/clock.go:48 +0x4d
  github.com/Netflix/spectator-go.TestHttpClient_PostJsonTimeout.func1()
      /home/r2/code/ext/spectator-go/http_client_test.go:108 +0x44
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2069 +0x51
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2887 +0xca
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1952 +0x87d

Goroutine 67 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1239 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1512 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1510 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1418 +0x3b3
  main.main()
      _testmain.go:135 +0x236

Goroutine 73 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3013 +0x644
  net/http/httptest.(*Server).goServe.func1()
      /usr/local/go/src/net/http/httptest/server.go:308 +0xd8
==================
--- FAIL: TestHttpClient_PostJsonTimeout (0.00s)
    testing.go:1093: race detected during execution of test
DEBUG: 2021/03/01 12:54:54 posting data to http://127.0.0.1:34955, payload 2 bytes
DEBUG: 2021/03/01 12:54:54 response HTTP 503:
DEBUG: 2021/03/01 12:54:54 posting data to http://127.0.0.1:34955, payload 2 bytes
DEBUG: 2021/03/01 12:54:54 response HTTP 503:
DEBUG: 2021/03/01 12:54:54 posting data to http://127.0.0.1:34955, payload 2 bytes
DEBUG: 2021/03/01 12:54:54 response HTTP 503:
INFO: 2021/03/01 12:54:55 registry config has no uri. Ignoring Start request
DEBUG: 2021/03/01 12:54:55 Got 1 measurements
DEBUG: 2021/03/01 12:54:55 Sending 1 measurements to http://127.0.0.1:38663
DEBUG: 2021/03/01 12:54:55 posting data to http://127.0.0.1:38663, payload 161 bytes
DEBUG: 2021/03/01 12:54:55 response HTTP 200:
DEBUG: 2021/03/01 12:54:55 Got 3 measurements
DEBUG: 2021/03/01 12:54:55 Sending 3 measurements to http://127.0.0.1:39485
DEBUG: 2021/03/01 12:54:55 posting data to http://127.0.0.1:39485, payload 248 bytes
DEBUG: 2021/03/01 12:54:55 response HTTP 202:
INFO: 2021/03/01 12:54:55 1 measurement(s) dropped due to validation errors: key too short: [o] (1 < 2)
DEBUG: 2021/03/01 12:54:55 Got 1 measurements
DEBUG: 2021/03/01 12:54:55 Sending 1 measurements to http://127.0.0.1:33923
DEBUG: 2021/03/01 12:54:55 posting data to http://127.0.0.1:33923, payload 161 bytes
DEBUG: 2021/03/01 12:54:55 response HTTP 200:
DEBUG: 2021/03/01 12:54:55 Got 6 measurements
DEBUG: 2021/03/01 12:54:55 Sending 6 measurements to http://127.0.0.1:33923
DEBUG: 2021/03/01 12:54:55 posting data to http://127.0.0.1:33923, payload 850 bytes
DEBUG: 2021/03/01 12:54:55 response HTTP 200:
DEBUG: 2021/03/01 12:54:55 Got 6 measurements
DEBUG: 2021/03/01 12:54:55 Got 5 measurements
DEBUG: 2021/03/01 12:54:55 Sending 5 measurements to http://127.0.0.1:33923
DEBUG: 2021/03/01 12:54:55 posting data to http://127.0.0.1:33923, payload 724 bytes
DEBUG: 2021/03/01 12:54:55 response HTTP 200:
FAIL
FAIL    github.com/Netflix/spectator-go 1.053s
ok      github.com/Netflix/spectator-go/histogram       0.159s
FAIL
```

Race example for `MapKey()` -- this is the test added in this changeset, but without the change to MapKey(). The scope of the test run is limited due to an unrelated test failure in `TestHttpClient_PostJsonTimeout` that is seemingly only triggered with the race detector enabled. i.e. this failure occurs on `master`.
```
➜ go test -race ./...  -run TestId
==================
WARNING: DATA RACE
Write at 0x00c00007efd8 by goroutine 9:
  github.com/Netflix/spectator-go.(*Id).MapKey()
      /home/r2/code/ext/spectator-go/id.go:55 +0x55a
  github.com/Netflix/spectator-go.TestId_mapKeyConcurrent.func1()
      /home/r2/code/ext/spectator-go/id_test.go:31 +0x38

Previous read at 0x00c00007efd8 by goroutine 10:
  github.com/Netflix/spectator-go.(*Id).MapKey()
      /home/r2/code/ext/spectator-go/id.go:20 +0x64
  github.com/Netflix/spectator-go.TestId_mapKeyConcurrent.func2()
      /home/r2/code/ext/spectator-go/id_test.go:35 +0x38

Goroutine 9 (running) created at:
  github.com/Netflix/spectator-go.TestId_mapKeyConcurrent()
      /home/r2/code/ext/spectator-go/id_test.go:30 +0xd8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202

Goroutine 10 (finished) created at:
  github.com/Netflix/spectator-go.TestId_mapKeyConcurrent()
      /home/r2/code/ext/spectator-go/id_test.go:34 +0x104
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
==================
--- FAIL: TestId_mapKeyConcurrent (0.00s)
    testing.go:1093: race detected during execution of test
FAIL
FAIL    github.com/Netflix/spectator-go 0.007s
ok      github.com/Netflix/spectator-go/histogram       (cached) [no tests to run]
FAIL
```

dmuino commented on clock.go:
```diff
@@ -28,6 +31,7 @@ func (c *SystemClock) Nanos() int64 {
 // source. You need to seed the time with either .SetFromDuration() or
 // .SetNanos().
 type ManualClock struct {
+	mu    sync.RWMutex
```
We should just use `sync/atomic` and do `LoadInt64(&nanos)/StoreInt64(&nanos, value)` to make it thread-safe. We don't need for a read-write mutex.

ralfonso replied:
Works for me. I've amended that commit and removed the updates to the tests since field order is no longer important.

dmuino reviewed (commented):
Looks good to me. The minor style issue with the `ManualClock` but this looks fine to merge with some caveats.

The reasoning behind making the map key a lazy value was to avoid having to compute it during times where we create temporary Ids which are never registered. A common use case is to have a base id (which is never added to a map, like the registry) but just used to create a few Ids `WithTag(...)` or `WithStat`/`WithDefaultStat`.  This is the case for all the `Measure()` functions which are now a bit more expensive, and we now might need to cache the `Id` used for them to avoid having to compute the key every time at the expense of using a bit more memory.

ralfonso commented:
@dmuino Thanks for the additional context on the lazy compute. 

Do you think a lock or a `sync.Once` would be better than the approach I've taken? I'm happy to amend to either.

dmuino commented:
> Do you think a lock or a `sync.Once` would be better than the approach I've taken? I'm happy to amend to either.

I think `sync.Once` is exactly what we need here to make it thread-safe. This way we do not need to worry about the temporary Ids.

It might be nice to actually cache the Ids use for `Measure` anyway to get a small performance improvement anyway but that is unrelated to this PR.

ralfonso commented:
@dmuino I've updated the implementation to use `sync.Once`. Please note there is a slight change in behavior from the previous lock-free implementation. See the commit message for more detail.

I've also added a commit to fix a couple of incorrect error messages that I discovered when running the tests locally.

brharrington commented on http_client_test.go:
```diff
@@ -86,7 +86,7 @@ func TestHttpClient_PostJson503_Compress(t *testing.T) {
 	})
 	if len(meters) != 3 {
-		t.Fatal("Expected 1 meter, got", len(meters))
+		t.Fatal("Expected 3 meter, got", len(meters))
```
`s/meter/meters/`

ralfonso replied:
Thanks, I've amended this commit.

brharrington reviewed (approved):
Not super familiar with Go, but the general change looks fine though as pointed out in the commit comment the key value for an error would be a bit odd.

Given the long term plan to delegate most of the logic to spectatord, I do wonder how worthwhile it is to have users investing in performance improvements and related bug fixes plus our time reviewing them for an implementation that should be short lived.

ralfonso commented:
> though as pointed out in the commit comment the key value for an error would be a bit odd.

@brharrington I may have been unclear. I did not introduce the string-as-error behavior in this change. I wanted to call out that in the previous implementation, an error encountered in `MapKey()` would have returned an `ERR` string and subsequent calls would have attempted to recompute. This change caches the `ERR` string and will not reattempt. 

It's possible to avoid the caching, but it will add additional complexity to this method and/or the use of a mutex. 

> Given the long term plan to delegate most of the logic to spectatord, I do wonder how worthwhile it is to have users investing in performance improvements and related bug fixes plus our time reviewing them for an implementation that should be short lived.

Understood, though I would argue that this particular method will still have utility when spectatord becomes the backend for this library as long as it retains a type for `Id`. 

+ @dmuino Thanks all for the review. Would it be possible to tag this as 1.1.2 after merge?
